metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestRetentionPoliciesStep.java

```diff
@@ -0,0 +1,144 @@
+
+  @Nonnull
+  @Override
+  public ExecutionMode getExecutionMode() {
+    return ExecutionMode.ASYNC;
+  }
+
+  @Override
+  public String name() {
+    return "IngestRetentionPoliciesStep";
+  }
+
+  @Override
+  public void execute() throws IOException, URISyntaxException {
+    // 0. Execute preflight check to see whether we need to ingest policies
+    log.info("Ingesting default retention...");
+
+    // Whether we are at clean boot or not.
+    if (_disableRetention) {
+      log.info("IngestRetentionPolicies disabled. Skipping.");
+      return;
+    }
+
+    // 1. Read default retention config
+    final Map<DataHubRetentionKey, DataHubRetentionInfo> retentionPolicyMap =
+        parseFileOrDir(new ClassPathResource("./boot/retention.yaml").getFile());
+
+    // 2. Read plugin retention config files from input path and overlay
+    retentionPolicyMap.putAll(parseFileOrDir(new File(pluginPath)));
+
+    // 4. Set the specified retention policies
+    log.info("Setting {} policies", retentionPolicyMap.size());
+    boolean hasUpdate = false;
+    for (DataHubRetentionKey key : retentionPolicyMap.keySet()) {
+      if (_retentionService.setRetention(key.getEntityName(), key.getAspectName(), retentionPolicyMap.get(key))) {
+        hasUpdate = true;
+      }
+    }
+
+    // 5. If there were updates on any of the retention policies, apply retention to all records
+    if (hasUpdate) {
+      log.info("Applying policies to all records");
+      _retentionService.batchApplyRetention(null, null);
+    }
+  }
+
+  // Parse input yaml file or yaml files in the input directory to generate a retention policy map
+  private Map<DataHubRetentionKey, DataHubRetentionInfo> parseFileOrDir(File retentionFileOrDir) throws IOException {
+    // If path does not exist return empty
+    if (!retentionFileOrDir.exists()) {
+      return Collections.emptyMap();
+    }
+
+    // If directory, parse the yaml files under the directory
+    if (retentionFileOrDir.isDirectory()) {
+      Map<DataHubRetentionKey, DataHubRetentionInfo> result = new HashMap<>();
+
+      for (File retentionFile : retentionFileOrDir.listFiles()) {
+        if (!retentionFile.isFile()) {
+          log.info("Element {} in plugin directory {} is not a file. Skipping", retentionFile.getPath(),
+              retentionFileOrDir.getPath());
+          continue;
+        }
+        result.putAll(parseFileOrDir(retentionFile));
+      }
+      return result;
+    }
+    // If file, parse the yaml file and return result;
+    if (!retentionFileOrDir.getPath().endsWith(".yaml") && retentionFileOrDir.getPath().endsWith(".yml")) {
+      log.info("File {} is not a YAML file. Skipping", retentionFileOrDir.getPath());
+      return Collections.emptyMap();
+    }
+    return parseYamlRetentionConfig(retentionFileOrDir);
+  }
+
+  private Map<DataHubRetentionKey, DataHubRetentionInfo> parseYamlRetentionConfig(File retentionConfigFile)
```

Comment:
this is great ! is there a comment about what the expected structure of a given yaml file should be?

Reply:
Let me add